### ISSUE TITLE
FacebookTemplateRenderer does not handle native banner format

### ISSUE THREAD

**Issue by JeremyR34** [mopub/mopub-android-mediation#242]
1. **What is the issue?**
Facebook native banner format is not correctly handled by FacebookTemplateRenderer, only standard native ad is; resulting by empty ad view.

Other info:
- I use manual integration with MoPubNative
- Same ad unit works with FacebookAdRenderer
- [Native FAN implementation](https://developers.facebook.com/docs/audience-network/guides/ad-formats/native/android-template) on this native banner works

2. **Check the applicable box(es) below by placing an `x` in the brackets:**
    - [x] I searched https://developers.mopub.com/ and found no answer.
    - [x] I searched https://twittercommunity.com/c/advertiser-api/mopub to make sure that this issue has not already been filed.
    - [x] I searched the adapter changelogs to make sure this issue has not already been resolved.

3. **Fill in the below template as detailed as possible:**
    * **Ad Format** (*Banner, Interstitial, Rewarded Video, Native Static, Native Video*)
        > Native Static - FAN Native banner
    * **Platform** (*Android, iOS, Unity*)
        > Android
    * **MoPub SDK Version**
        > 5.11.1
    * **Mediation SDK Name & Version**
        > audience-network-sdk 5.7.1
    * **Mediation Adapter Version**
        > 5.7.1.1
    * **MoPub Ad Unit ID(s) used to reproduce the issue**
        > NA
    * **Device model and OS Version**
        > NA
    * **Steps to reproduce the issue** (include integration steps if not standard)
        > 1. Use FacebookTemplateRenderer as described in the [documentation](https://developers.mopub.com/publishers/android/integrate-networks/) with a Facebook Native banner unit
    * **Observed behavior**
        > Empty view returned by getAdView() when rendering the ad.
    * **Expected behavior** (if known)
        > Render an ad view according to the provided template parameters
    * **Evidence** (if available)
        > 
In FacebookTemplateRenderer code a check is done on nativeAdBase class. In Native Banner case class is NativeBannerAd not NativeAd so the rendering is never done.
https://github.com/mopub/mopub-android-mediation/blob/9399704881b8d780fb2e6b8f25647e7950ac3451/FacebookAudienceNetwork/src/main/java/com/mopub/nativeads/FacebookTemplateRenderer.java#L42

**Comment by chauduyphanvu** [comment]
@JeremyR34, could you provide runnable code (along with ad unit IDs) that reproduces this issue? Thanks!

Vu Chau
MoPub Team

**Comment by JeremyR34** [comment]
Yes sure.

```java
protected void loadMoPubNativeAd() {
      MoPubNative moPubNative = new MoPubNative(this, "e94e77376f0b43948dd5e563a06bc4c3", new MoPubNative.MoPubNativeNetworkListener() {
        @Override
        public void onNativeLoad(final NativeAd nativeAd) {
          AdapterHelper adapterHelper = new AdapterHelper(MyActivity.this, 0, 3); // When standalone, any range will be fine.
          // Render ad view
          View adView = adapterHelper.getAdView(null, null, nativeAd, new ViewBinder.Builder(0).build());
          // Add view to layout
          addMoPubNativeViewToLayout(adView); // adView is emty here
        }

        @Override
        public void onNativeFail(NativeErrorCode errorCode) {

        }
      });

      ViewBinder viewBinder = new ViewBinder.Builder(R.layout.ad_native_mopub)
            .iconImageId(R.id.appinstall_app_icon)
            .titleId(R.id.appinstall_headline)
            .textId(R.id.appinstall_body)
            .privacyInformationIconImageId(R.id.image_privacy)
            .callToActionId(R.id.appinstall_call_to_action)
            .addExtra("starrating", R.id.appinstall_stars)
            .build();

      FacebookTemplateRenderer facebookTemplateRenderer = null;
      facebookTemplateRenderer = new FacebookTemplateRenderer(new NativeAdViewAttributes());

      /* Facebook ad renderer that works fine
      FacebookAdRenderer facebookAdRenderer = new FacebookAdRenderer(
          new FacebookAdRenderer.FacebookViewBinder.Builder(R.layout.ad_native_facebook)
              .textId(R.id.appinstall_body)
              // Binding to new layouts from Facebook 4.99.0+
              .adIconViewId(R.id.appinstall_app_icon)
              .adChoicesRelativeLayoutId(R.id.ad_choices_container)
              .advertiserNameId(R.id.appinstall_headline) // Bind either the titleId or advertiserNameId depending on the FB SDK version
              // End of binding to new layouts
              .callToActionId(R.id.appinstall_call_to_action)
              .sponsoredNameId(R.id.native_ad_sponsored_label)
              .build());*/

      EnumSet<RequestParameters.NativeAdAsset> desiredAssets = EnumSet.of(
          RequestParameters.NativeAdAsset.TITLE,
          RequestParameters.NativeAdAsset.TEXT,
          RequestParameters.NativeAdAsset.CALL_TO_ACTION_TEXT,
          RequestParameters.NativeAdAsset.ICON_IMAGE,
          RequestParameters.NativeAdAsset.STAR_RATING,
          RequestParameters.NativeAdAsset.SPONSORED
      );

      moPubNative.registerAdRenderer(facebookTemplateRenderer);

      MoPubStaticNativeAdRenderer moPubStaticNativeAdRenderer = new MoPubStaticNativeAdRenderer(viewBinder);
      moPubNative.registerAdRenderer(moPubStaticNativeAdRenderer); // MoPub should be last

      RequestParameters mRequestParameters = new RequestParameters.Builder()
          .desiredAssets(desiredAssets)
          .build();
      // Load ad
      Map<String, Object> localExtras = new HashMap<>();
      localExtras.put("native_banner", true); // For facebook native banner
      moPubNative.setLocalExtras(localExtras);
      moPubNative.makeRequest(mRequestParameters);
  }
```

**Comment by chauduyphanvu** [comment]
@JeremyR34, thanks! We've verified the behavior you mentioned, and will work with Facebook on supporting this use case. Stay tuned!

**Comment by chauduyphanvu** [comment]
@JeremyR34, thanks for the patience. This is now resolved in https://github.com/mopub/mopub-android-mediation/commit/fabdcf9ec05cb53d8a5c5c8b4755182e20573bb2.

To use native banner via the Facebook template:

```java

FacebookTemplateRenderer facebookTemplateRenderer = new FacebookTemplateRenderer(new NativeAdViewAttributes(getContext()), NativeBannerAdView.Type.HEIGHT_120);

mMoPubNative.registerAdRenderer(facebookTemplateRenderer);

Map<String, Object> localExtras = new HashMap<>();
localExtras.put("native_banner", true);
mMoPubNative.setLocalExtras(localExtras);

// Make your ad request
```

The `FacebookTemplateRenderer` constructor accepts a non-null `NativeBannerAdView.Type` param. Let us know if you need anything else!

**Comment by hoanghiephui** [comment]
> @JeremyR34, thanks for the patience. This is now resolved in [fabdcf9](https://github.com/mopub/mopub-android-mediation/commit/fabdcf9ec05cb53d8a5c5c8b4755182e20573bb2).
> 
> To use native banner via the Facebook template:
> 
> ```java
> FacebookTemplateRenderer facebookTemplateRenderer = new FacebookTemplateRenderer(new NativeAdViewAttributes(getContext()), NativeBannerAdView.Type.HEIGHT_120);
> 
> mMoPubNative.registerAdRenderer(facebookTemplateRenderer);
> 
> Map<String, Object> localExtras = new HashMap<>();
> localExtras.put("native_banner", true);
> mMoPubNative.setLocalExtras(localExtras);
> 
> // Make your ad request
> ```
> 
> The `FacebookTemplateRenderer` constructor accepts a non-null `NativeBannerAdView.Type` param. Let us know if you need anything else!

Hi @chauduyphanvu 
Can you help me. I was add FacebookTemplateRenderer. But I have an ad width not being filled

**Comment by chauduyphanvu** [comment]
@hoanghiephui, since this issue is old, could you create a separate issue on GitHub and provide the details specific to your integration? We should then be able to assist!

Thanks.
